src/wallet/sqlite.cpp

```diff
@@ -606,10 +612,13 @@ std::unique_ptr<DatabaseCursor> SQLiteBatch::GetNewPrefixCursor(Span<const std::
 
 bool SQLiteBatch::TxnBegin()
 {
+    m_database.m_sqlite_semaphore.wait();
     if (!m_database.m_db || sqlite3_get_autocommit(m_database.m_db) == 0) return false;
     int res = sqlite3_exec(m_database.m_db, "BEGIN TRANSACTION", nullptr, nullptr, nullptr);
     if (res != SQLITE_OK) {
         LogPrintf("SQLiteBatch: Failed to begin the transaction\n");
+    } else {
+        m_txn = true;
```

Comment:
This can also cause a deadlock when the statement execution fails.

The semaphore is not released after a `TxnBegin` failure which will block forever if the software continues and executes a single write or another `TxnBegin` attempt. Which could happen as `TxnBegin` failure could be temporary.

Reply:
Fixed